src/tests/math.rs

```diff
@@ -0,0 +1,142 @@
+#[test_case("$22 and $$2+2$$", "<p>$22 and <math>2+2</math></p>\n")]
+#[test_case("$$a!$$", "<p><math>a!</math></p>\n")]
+#[test_case("$$x$$", "<p><math>x</math></p>\n")]
+#[test_case("$$20,000 and $$30,000", "<p><math>20,000 and </math>30,000</p>\n")]
+#[test_case(
+    "$$22+1$$ and $$22 + a^2$$",
+    "<p><math>22+1</math> and <math>22 + a^2</math></p>\n"
+)]
+#[test_case(
+    "$$2+2$$ $22 and dollars$22 $$2+2$$",
+    "<p><math>2+2</math> $22 and dollars$22 <math>2+2</math></p>\n"
+)]
+#[test_case(
+    "dollars$22 and $$a^2 + b^2 = c^2$$",
+    "<p>dollars$22 and <math>a^2 + b^2 = c^2</math></p>\n"
+)]
+fn math_dollars_inline_display(markdown: &str, html: &str) {
+    let result = html
+        .replace("<math>", "<span data-math-style=\"display\">")
+        .replace("</math>", "</span>");
+
+    html_opts!([extension.math_dollars], markdown, &result);
+}
+
+#[test_case("$$\n2+2\n4+4\n$$", "<p><math>2+2\n4+4\n</math></p>\n")]
+fn math_dollars_block(markdown: &str, html: &str) {
+    let result = html
+        .replace("<math>", "<span data-math-style=\"display\">")
+        .replace("</math>", "</span>");
+
+    html_opts!([extension.math_dollars], markdown, &result);
+}
+
+#[test_case("$`2+2`$", "<p><math>2+2</math></p>\n")]
+#[test_case("$22 and $`2+2`$", "<p>$22 and <math>2+2</math></p>\n")]
+#[test_case("$`1+\\$2`$", "<p><math>1+\\$2</math></p>\n")]
+#[test_case(
+    "$`22+1`$ and $`22 + a^2`$",
+    "<p><math>22+1</math> and <math>22 + a^2</math></p>\n"
+)]
+#[test_case(
+    "$`2+2`$ $22 and dollars$22 $`2+2`$",
+    "<p><math>2+2</math> $22 and dollars$22 <math>2+2</math></p>\n"
+)]
+fn math_code_inline(markdown: &str, html: &str) {
+    let result = html
+        .replace("<math>", "<code data-math-style=\"inline\">")
+        .replace("</math>", "</code>");
+
+    html_opts!([extension.math_code], markdown, &result);
+}
+
+#[test_case("```math\n2+2\n4+4\n```", "<pre><math>2+2\n4+4\n</math></pre>\n")]
+fn math_code_block(markdown: &str, html: &str) {
+    let result = html
+        .replace(
+            "<math>",
+            "<code class=\"language-math\" data-math-style=\"display\">",
+        )
+        .replace("</math>", "</code>");
+
+    html_opts!([extension.math_dollars], markdown, &result);
+}
+
+#[test_case("`2+2`", "<p><code>2+2</code></p>\n")]
+#[test_case("test `2+2`$ test", "<p>test <code>2+2</code>$ test</p>\n")]
+#[test_case("$20,000 and $30,000", "<p>$20,000 and $30,000</p>\n")]
+#[test_case("$20,000 in $USD", "<p>$20,000 in $USD</p>\n")]
+#[test_case("$ a^2 $", "<p>$ a^2 $</p>\n")]
+#[test_case("$\n$", "<p>$\n$</p>\n")]
+#[test_case("$$$", "<p>$$$</p>\n")]
+#[test_case("`$1+2$`", "<p><code>$1+2$</code></p>\n")]
+#[test_case("`$$1+2$$`", "<p><code>$$1+2$$</code></p>\n")]
+#[test_case("`$\\$1+2$$`", "<p><code>$\\$1+2$$</code></p>\n")]
+// #[test_case("test $`2+2` test", "<p>test $<code>2+2</code> test</p>\n")]
+// #[test_case("test $$\n2+2\n$$", "<p>test $$\n2+2\n$$</p>\n")]
```

Comment:
The results of these tests at present are interesting:

```html
<p>test $2+2` test</p>
```

```html
<p>test $$
2+2</p>
<p><span data-math-style="display"></span></p>
```

I think the stated (commented-out) expectations are good; any thoughts here?

Reply:
I was able to fix a few things with this. Ended up being about to completely remove the `NodeMathBlock` and any related code. Since we're wrapping 

```
$$
1+2
$$
```

in paragraph tags, rather than code tags, then inline parsing of dollars handles all the cases. It reduced the complexity of the code, and brings it much closer to https://github.com/jgm/commonmark-hs/blob/master/commonmark-extensions/test/math.md